### PR TITLE
HW11: implemented Transactional annotation, added loading from a file, reworked DI

### PR DESCRIPTION
**Pull request by my20yeezy**

### CODE REVIEW

File: src/main/java/com/ernie/TicketApp/ApplicationContext.java [hunk 1]
```diff
@@ -25,6 +32,7 @@ private static void loadProperties() {
         Properties properties = new Properties();
         try (FileInputStream input = new FileInputStream("src/main/resources/application.properties")) {
             properties.load(input);
+            dbClassName = properties.getProperty("db.className");
             dbUrl = properties.getProperty("db.url");
             dbUsername = properties.getProperty("db.username");
             dbPassword = properties.getProperty("db.password");
```
Comment (AtticSiol) [inline src/main/java/com/ernie/TicketApp/ApplicationContext.java]: do not load properties with the help of `FileInputStream` and ` Properties`. Use 1 of `Spring` approaches:
1. add `@PropertySource("classpath:application.properties")`
   and read values with `@Value`:
```java
 @Value("${db.url}")
 private String dbUrl;
   ... 
    dataSource.setUrl(dbUrl);
   ...
 ```

2. add `@PropertySource("classpath:application.properties")`
   and read values with `Spring Environment`:
```java
private final Environment env;
public ApplicationContext(Environment env) {this.env = env;}
   ... 
    dataSource.setUrl("db.url"));
   ...
 ```

File: src/main/java/com/ernie/TicketApp/service/TicketService.java [hunk 2]
```diff
@@ -1,138 +1,21 @@
...
-        myHashSet.print();
-
-        System.out.println(myHashSet.contains("Element 4"));
-        System.out.println(myHashSet.contains("Element 2"));
-
-        myHashSet.delete("Element 1");
 
-        myHashSet.print();
+    @Value("${user_ticket_access.enabled}")
+    public boolean isEnabled;
```
Comment (AtticSiol) [inline src/main/java/com/ernie/TicketApp/service/TicketService.java]: use full name, for ex.: `isUserTicketAccessEnabled`

File: src/main/java/com/ernie/TicketApp/repository/UserDAO.java [hunk 3]
```diff
@@ -93,4 +95,46 @@ public List<User> getAllUsers() {
         return allUsers;
     }
 
+    @Transactional
+    public void updateUserAddTicket(UUID id) {
```
Comment (AtticSiol) [inline src/main/java/com/ernie/TicketApp/repository/UserDAO.java:102]: the core idea of p.6 from HW11 is correct, but the implementation - not. Follow next instructions:
1. when you work with  `try (PreparedStatement preparedStatement = dataSource.getConnection().prepareStatement(...)) {...}` then:

-  you are responsible for all interactions with database, while the idea is to transfer this responsibility to Spring.
- each call of `executeUpdate()` method will make changes in database independently, while the idea is to perform them in 1 transaction. Both of 2nd and 3d changes have to be performed or rolled back

In order to fix it, you can:
- work with `JdbcTemplate` + `@Transactional` instead of `try (PreparedStatement preparedStatement = dataSource.getConnection().prepareStatement(...)) {...}` :
add this bean in configuration class:
```java
    @Bean
    public JdbcTemplate jdbcTemplate(DataSource dataSource) {
        return new JdbcTemplate(dataSource);
    }
```
and to `UserDAO` class:
```java
    private final JdbcTemplate jdbcTemplate;
    public UserDAO(DataSource dataSource, JdbcTemplate jdbcTemplate) {
        this.dataSource = dataSource;
        this.jdbcTemplate = jdbcTemplate;
    }
```
- call 2 queries on this object:
```java
jdbcTemplate.update("your query", newTicket.getId().toString(), user.getId().toString(), newTicket.getTicketType().name(), newTicket.getCreationDateTime().toString());
jdbcTemplate.update("your query", yourVariables);
```

File: src/main/java/com/ernie/TicketApp/repository/UserDAO.java [hunk 4]
```diff
@@ -93,4 +95,46 @@ public List<User> getAllUsers() {
...
+            ResultSet resultSet = preparedStatement.executeQuery();
+            if (resultSet.next()) {
+                String name = resultSet.getString("name");
+                user = new User(name);
+                user.setId(UUID.fromString(resultSet.getString("id")));
+                user.setCreationDateTime(LocalDateTime.parse(resultSet.getString("creation_date")));
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
```
Comment (AtticSiol) [inline src/main/java/com/ernie/TicketApp/repository/UserDAO.java]: this logic is not related to ` update User and create ticket ` task. 
move on to service class:
```java
        Ticket newTicket = new Ticket();
        newTicket.setTicketType(TicketType.MONTH);
        User user = userDAO.getUserById(...); // retrieve user
        userDAO.updateUserAddTicket(user, ....); //  update User and create ticket
```

**Comment by my20yeezy** [comment]
@AtticSiol  fixed all in https://github.com/my20yeezy/Learning-at-Andersen/pull/9/commits/4257e142daa2da27cc42a963a336badad9d490dd

File: src/main/java/com/ernie/TicketApp/repository/UserDAO.java [hunk 5]
```diff
@@ -93,4 +98,17 @@ public List<User> getAllUsers() {
         return allUsers;
     }
 
+    @Transactional
+    public void updateUserAddTicket(UUID id) {
+        Ticket newTicket = new Ticket();
+        newTicket.setTicketType(TicketType.MONTH);
```
Comment (AtticSiol) [inline src/main/java/com/ernie/TicketApp/repository/UserDAO.java:104]: this is service layer logic. all data has to be already prepared, when you work with it on dao layer

**Comment by AtticSiol** [comment]
approved. 1 comment just for your attention. you can merge this branch to main

**Review by AtticSiol — APPROVED** [review]
